lib/auth/keystore/gcp_kms.go

```diff
@@ -0,0 +1,454 @@
+		"SOFTWARE": kmspb.ProtectionLevel_SOFTWARE,
+		"HSM":      kmspb.ProtectionLevel_HSM,
+	}
+)
+
+// GCPKMS holds configuration parameters specific to GCP KMS keystores.
+type GCPKMSConfig struct {
+	// KeyRing is the fully qualified name of the GCP KMS keyring.
+	KeyRing string
+	// ProtectionLevel specifies how cryptographic operations are performed.
+	// For more information, see https://cloud.google.com/kms/docs/algorithms#protection_levels
+	// Supported options are "hsm" and "software".
+	ProtectionLevel string
+	// HostUUID is the UUID of the teleport host (auth server) running this
+	// keystore. Used to label keys so that they can be queried and deleted per
+	// server without races when multiple auth servers are configured with the
+	// same KeyRing.
+	HostUUID string
+
+	kmsClientOverride kmsClientI
+	clockOverride     clockwork.Clock
+}
+
+func (cfg *GCPKMSConfig) CheckAndSetDefaults() error {
+	if cfg.KeyRing == "" {
+		return trace.BadParameter("must provide a valid KeyRing to GCPKMSConfig")
+	}
+	if _, ok := gcpKMSProtectionLevels[cfg.ProtectionLevel]; !ok {
+		return trace.BadParameter("unsupported ProtectionLevel %s", cfg.ProtectionLevel)
+	}
+	if cfg.HostUUID == "" {
+		return trace.BadParameter("must provide a valid HostUUID to GCPKMSConfig")
+	}
+	return nil
+}
+
+type gcpKMSKeyStore struct {
+	hostUUID        string
+	keyRing         string
+	protectionLevel kmspb.ProtectionLevel
+	kmsClient       kmsClientI
+	log             logrus.FieldLogger
+	clock           clockwork.Clock
+}
+
+// newGCPKMSKeyStore returns a new keystore configured to use a GCP KMS keyring
+// to manage all key material.
+func newGCPKMSKeyStore(cfg *GCPKMSConfig) (*gcpKMSKeyStore, error) {
+	var kmsClient kmsClientI
+	if cfg.kmsClientOverride != nil {
+		kmsClient = cfg.kmsClientOverride
+	} else {
+		realClient, err := kms.NewKeyManagementClient(context.Background())
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		kmsClient = &kmsClientWrapper{realClient}
+	}
+
+	clock := clockwork.NewRealClock()
+	if cfg.clockOverride != nil {
+		clock = cfg.clockOverride
+	}
+
+	return &gcpKMSKeyStore{
+		hostUUID:        cfg.HostUUID,
+		keyRing:         cfg.KeyRing,
+		protectionLevel: gcpKMSProtectionLevels[cfg.ProtectionLevel],
+		kmsClient:       kmsClient,
+		log:             logrus.WithFields(logrus.Fields{trace.Component: "GCPKMSKeyStore"}),
+		clock:           clock,
+	}, nil
+}
+
+func newKeyID() (string, error) {
+	id, err := uuid.NewRandom()
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+	return id.String(), nil
+}
+
+// generateRSA creates a new RSA private key and returns its identifier and a
+// crypto.Signer. The returned identifier for gcpKMSKeyStore encoded the full
+// GCP KMS key version name, and can be passed to getSigner later to get the same
+// crypto.Signer.
+func (g *gcpKMSKeyStore) generateRSA(opts ...RSAKeyOption) ([]byte, crypto.Signer, error) {
+	ctx := context.TODO()
```

Comment:
Why not take a `context`?

Reply:
because it's a 300 line diff :( I meant to make this a legitimate TODO, follow-up PR here https://github.com/gravitational/teleport/pull/18034